domain/relation/state/relation.go

```diff
@@ -481,6 +483,389 @@ WHERE  relation_id = $getRelation.relation_id
+		return "", relationerrors.UnitNotFound
+	} else if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	return args.ApplicationUUID, nil
+}
+
+// getApplicationsInRelation gets all the applications that are in the given
+// relation.
+func (st *State) getApplicationsInRelation(
+	ctx context.Context,
+	tx *sqlair.TX,
+	uuid corerelation.UUID,
+) ([]application.ID, error) {
+	type applicationUUID struct {
+		UUID application.ID `db:"application_uuid"`
+	}
+	relUUID := relationUUID{
+		UUID: uuid.String(),
+	}
+	stmt, err := st.Prepare(`
+SELECT &applicationUUID.application_uuid
+FROM   relation_endpoint re 
+JOIN   application_endpoint ae ON re.endpoint_uuid = ae.uuid
+WHERE  re.relation_uuid = $relationUUID.uuid
+`, relUUID, applicationUUID{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	var apps []applicationUUID
+	err = tx.Query(ctx, stmt, relUUID).GetAll(&apps)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return nil, relationerrors.RelationNotFound
+	} else if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var ids []application.ID
+	for _, app := range apps {
+		ids = append(ids, app.UUID)
+	}
+
+	return ids, nil
+}
+
+// checkSubordinateUnitCanEnterScope checks that subordinate units can enter
+// scope. Under the following conditions, a subordinate unit cannot enter a
+// relation that is not with its principle application:
+//   - The relation has scope container.
+//   - The unit entering scope is a subordinate.
+//   - The other application is not a subordinate.
+func (st *State) checkSubordinateUnitCanEnterScope(
+	ctx context.Context,
+	tx *sqlair.TX,
+	relUUID corerelation.UUID,
+	unitUUID unit.UUID,
+	otherApplication application.ID,
+) error {
+	// Check that the relation is a container scoped relation.
+	scope, err := st.getRelationScope(ctx, tx, relUUID)
+	if err != nil {
+		return errors.Errorf("getting relation scope: %w", err)
+	}
+	if scope != charm.ScopeContainer {
+		return nil
+	}
+
+	// Check that the other application in the relation is not a subordinate
+	// application, if it is, we have a relation between two subordinates, which
+	// is OK.
+	if subordinate, err := st.isSubordinate(ctx, tx, otherApplication); err != nil {
+		return errors.Errorf("checking if application is subordinate: %w", err)
+	} else if !subordinate {
+		return nil
+	}
+
+	// Check that the principle application of the unit is the other application
+	// in the relation the unit is trying to enter scope in.
+	principleAppID, err := st.getPrincipalApplicationOfUnit(ctx, tx, unitUUID)
+	if err != nil {
+		return errors.Errorf("getting principal application of unit: %w", err)
+	}
+	if principleAppID != otherApplication {
+		return errors.Errorf(
+			"unit cannot enter scope: principle application not in relation: %w",
+			relationerrors.PotentialRelationUnitNotValid,
+		)
+	}
+
+	return nil
+}
+
+// getRelationScope returns the scope of the given relation.
+func (st *State) getRelationScope(
```

Comment:
question: where is the guarantee which both endpoints of the relation have the same scope? 

You're getting one of two possible values here for the relation, unless it's a peer relation.

Reply:
I've done some investigation into this. You can relate a charm defined Container scoped endpoint to a charm defined Global endpoint, but actually, [the AddRelation method changes the endpoint scope if either of them are Container](https://github.com/juju/juju/blob/c6c9fd4b8aaf9cc3e0b46df4017856881d1c6fe2/state/state.go#L2128-L2138). This means the application endpoint scope can actually differ from the relation endpoint scope, which is not something we currently have modeled explicitly.

For now, I have changed the implementation of this function to read out the relation scope as what the matching relation endpoints scopes would be in 3.6. This happens to be the same as @gfouillet's intuitive definition.

I've also put more explanative comments on the function.